### PR TITLE
Implement hooks

### PR DESCRIPTION
**Pull request by toddbaert**

### CODE REVIEW

File: packages/api/src/main.ts [hunk 1]
```diff
@@ -6,18 +6,50 @@ import * as express from 'express';
 import { openfeature } from '@openfeature/openfeature-js';
 import { fibonacci } from '@openfeature/fibonacci';
 import { query, validationResult } from 'express-validator';
+import { OpenTelemetryHook, LoggingHook } from '@openfeature/extra';
+import { buildMarkup } from './markup';
 
 const app = express();
 
-const oFeatClient = openfeature.getClient('api');
+openfeature.registerHooks(new OpenTelemetryHook('test'));
```
Comment (beeme1mr) [inline packages/api/src/main.ts:14]: Give the OTel hook a better name. It's supposed to uniquely identify the component which generated the instrumentation.

File: packages/js-cloudbees-provider/src/lib/js-cloudbees-provider.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import {
   Context,
-  FeatureProvider, parseValidJsonObject
+  FeatureProvider, FlagEvaluationOptions, parseValidJsonObject
```
Comment (beeme1mr) [inline packages/js-cloudbees-provider/src/lib/js-cloudbees-provider.ts:3]: Format with prettier

File: packages/js-env-provider/src/lib/js-env-provider.ts [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import {
   Context,
-  FeatureProvider, FlagTypeError,
+  FeatureProvider, FlagEvaluationOptions, FlagTypeError,
```
Comment (beeme1mr) [inline packages/js-env-provider/src/lib/js-env-provider.ts:3]: Format with prettier

File: packages/js-split-provider/src/lib/js-split-provider.ts [hunk 4]
```diff
@@ -21,17 +21,18 @@ export class OpenFeatureSplitProvider implements FeatureProvider {
...
 
   /**
    * Split doesn't directly handle booleans as treatment values.
    * It will be up to the provider author and it's users to come up with conventions for converting strings to booleans.
    */
-  async getBooleanValue(flagId: string, defaultValue: boolean, context?: Context): Promise<boolean> {
+  async getBooleanValue(flagId: string, defaultValue: boolean, context: Context, options?: FlagEvaluationOptions): Promise<boolean> {
     await this.initialized;
-    const stringValue = this.client.getTreatment(context?.userId ?? 'anonymous', flagId);
+    // simply casting Context to Attributes is likely a bad idea.
```
Comment (beeme1mr) [inline packages/js-split-provider/src/lib/js-split-provider.ts:34]: Will this be handled in a follow up PR?

File: packages/openfeature-js/src/lib/client.ts [hunk 5]
```diff
@@ -1,106 +1,213 @@
...
+  ): Promise<T> {
+    const provider = this.getProvider();
+    const flagHooks = options?.hooks || [];
+    const allHooks: Hook<FlagValue>[] = [...OpenFeatureAPI.getInstance().hooks , ...this.hooks, ...flagHooks];
+    context = context || {};
+    let hookContext: HookContext = {
+      flagId,
+      flagType,
+      defaultValue,
+      context: context || {},
```
Comment (beeme1mr) [inline packages/openfeature-js/src/lib/client.ts:92]: This is already handled on line 87.

File: packages/openfeature-js/src/lib/client.ts [hunk 6]
```diff
@@ -1,106 +1,213 @@
...
+        return hook.error(hookContext, error);
+      }
     });
   }
 
   private getProvider(): FeatureProvider {
     return this.api.getProvider() ?? NOOP_FEATURE_PROVIDER;
   }
+
+  private isError(err: unknown): err is Error {
```
Comment (beeme1mr) [inline packages/openfeature-js/src/lib/client.ts:208]: Is this used?

File: packages/openfeature-js/src/lib/types.ts [hunk 7]
```diff
@@ -1,53 +1,99 @@
...
 
-// eslint-disable-next-line @typescript-eslint/no-empty-interface
-export interface Client extends Features {}
+// consider this name.
+export interface HasHooks {
+  registerHooks(...hooks: Hook[]): void;
+  get hooks(): Hook[];
+}
+
+// may want to convert this to an interface later.
```
Comment (beeme1mr) [inline packages/openfeature-js/src/lib/types.ts:77]: What do you mean by this comment?

Comment (toddbaert) [reply packages/openfeature-js/src/lib/types.ts:77]: It's from when it was an abstract class.

File: packages/openfeature-js/src/lib/client.ts (hunk 6 above)
Comment (toddbaert) [reply packages/openfeature-js/src/lib/client.ts:208]: Yes, line 151 - I think you wrote this method

File: packages/js-split-provider/src/lib/js-split-provider.ts (hunk 4 above)
Comment (toddbaert) [reply packages/js-split-provider/src/lib/js-split-provider.ts:34]: Ya, was going to handle it with some of the upcoming Context research.

File: packages/openfeature-js/src/lib/client.ts [hunk 8]
```diff
@@ -1,106 +1,237 @@
...
         case 'json': {
-          valuePromise = provider.getObjectValue(id, defaultValue as object, context);
+          valuePromise = provider.getObjectValue(
+            flagId,
+            defaultValue as object,
+            context,
+            options
+          );
           break;
         }
```
Comment (beeme1mr) [inline packages/openfeature-js/src/lib/client.ts:158]: Perhaps we should add a default case that throws.

Comment (toddbaert) [reply packages/openfeature-js/src/lib/client.ts:158]: I can't see how that would hurt, in the event of some strange typo.

**Review by beeme1mr — APPROVED** [review]

**Pull request merged**